File: list/list.go
package main

import "fmt"

type Cons struct {
  Head int
  Tail *Cons
}

func Get(n int, list *Cons) interface {} {
  if list == nil {
    return nil
  } else if n == 0 {
    return list.Head
  } else {
    return Get(n - 1, list.Tail)
  }
}

func Size(list *Cons) int {
  if list == nil {
    return 0
  } else {
    return 1 + Size(list.Tail)
  }
}

func main() {
  foo := &Cons{0, &Cons{1, &Cons{2, nil}}}
  fmt.Println(Size(foo) == 3)
  fmt.Println(Get(1, foo) == 1)
  fmt.Println(Get(2, foo) == 2)
  fmt.Println(Get(3, foo) == nil)
}

File: list/list.pl
use Class::Struct;

struct Cons => {
    head => '$',
    tail => 'Cons'
};

sub cons {
    my ($head, $tail) = @_;
    my $cons = new Cons;
    $cons->head($head);
    if ($tail) {
	$cons->tail($tail);
    }
    return $cons;
}

sub size {
    my ($list) = @_;
    if ($list) {
	return 1 + size($list->tail());
    } else {
	return 0;
    }
}

sub get {
    my ($n, $list) = @_;
    if (!$list) {
	return $list;
    } elsif ($n == 0) {
	return $list->head();
    } else {
	return get($n - 1, $list->tail());
    }
}

my $foo = cons(0, cons(1, cons(2)));
print size($foo) == 3, "\n";
print get(1, $foo) == 1, "\n";
print get(2, $foo) == 2, "\n";
print get(3, $foo) == (), "\n";
